### ISSUE TITLE
Adding TAU NIGENS SSE 2021

### ISSUE THREAD

**Issue by iranroman** [soundata/soundata#57]
I'm working on adding the TAU NIGENS SSE 2021 dataset (DCASE Task 3 2021). 

The data consists of 1-min recordings with spatial events and annotations in a csv file. The annotations have these columns for each event

`time-frame number`, `active class index`, `event number index`, `azimuth`, and `elevation` (all integers)

* `time-frame` is a timestamp
* `active class index` is the class ID of the event
* `event number index` is a sequence of increasing integers to indicate the order of the events and avoid confusion between overlapping events of the same class.
* `azimuth` and `elevation` are self-explanatory (in units of degrees). 

A new annotation class is needed. I was thinking of creating a `SpatialEvents` class. What do you think?

This annotation class will also be useful for other datasets I will add:

* TAU NIGENS SSE 2020. Exact same annotations. 

* TAU SSE 2019. Annotations are `frame number`, `active class index`, `event number index`, `azimuth`, and `elevation`.

Azimuth and elevation are annotations we will keep running into when dealing with datasets capturing spatialized events. In TAU SSE 2019, the distance is also included. I suspect more datasets will also include that.

**Comment by magdalenafuentes** [comment]
Thanks for this @iranroman!

We discussed this offline a bit and I agree that a new `annotation type` might be needed. Gonna drop an issue about that specifically

**Comment by iranroman** [comment]
I think what we need is a `SpatialEvents` class that accounts for `azimuth`, `elevation`, and `distance` of events. This would be a simple `Events` class spinoff. 

```
class SpatialEvents(Annotation):
    """Events class

    Attributes:
        intervals (np.ndarray): (n x 2) array of intervals
            (as floats) in seconds in the form [start_time, end_time]
            with positive time stamps and end_time >= start_time.
        labels (list): list of event labels (as strings)
        azimuth (list): list of azimuth degree (as int in the rante [0 360])
        elevation (list): list of elevation degree (as int in the range [0, 180])
        distance (list): list of distance in meters (as float, or None)
        confidence (np.ndarray or None): array of float confidence values
            in the range [0, 1]
    """

    def __init__(self, intervals, labels, azimuth, elevation, distance, confidence=None) -> None:
        validate_array_like(intervals, np.ndarray, float)
        validate_array_like(labels, list, str)
        validate_array_like(azimuth, list, int)
        validate_array_like(elevation, list, int)
        validate_array_like(distance, list, float, none_allowed=True)
        validate_array_like(confidence, np.ndarray, float, none_allowed=True)
        validate_lengths_equal([intervals, labels, azimuth, elevation, distance, confidence])
        validate_intervals(intervals)
        validate_confidence(confidence)
        validate_azimuth(azimuth)
        validate_elevation(elevation)

        self.intervals = intervals
        self.labels = labels
        self.confidence = confidence
        self.confidence = azimuth
        self.elevation = elevation
        self.distance = distance
```

For the specific case of TAU NIGENS 2020 and 2021, those dataset modules can define their own `IndexedSpatialEvents` class that inherits from the `SpatialEvents` class and accounts for the `event number index`, which I'm not sure will be broadly relevant for more datasets.

**Comment by justinsalamon** [comment]
Sorry I'm late to the discussion, I'll look for the other thread @magdalenafuentes created.

My 2c is that we should think hard before adding a new annotation type - could this be folded into the existing events, for example? Let's discuss here: #58

**Comment by iranroman** [comment]
Hello again, 

Thanks for your suggestions. I've been making significant progress on the loader. I have a question and want to hear your thoughts though.

The soundata `Events` class uses the start and end timestamps of the event to identify their location in time and duration. However, when it comes to the `TAU NIGENS SSE 2021` dataset, some events do not have such a clear-cut start and end. Let me explain: while the annotations have an absolute start and end of the event, an event may disappear momentarily in between, sometimes for longer than 1 second. This makes it tricky to determine what is the beginning and the end of the event. 

The event classes in this dataset are `alarm, crying baby, crash, barking dog, female scream, female speech,footsteps, knocking on door, male scream, male speech, ringing phone, piano`. So for example, while the annotations precisely say when a specific baby starts and ends crying, in between this window of time the same baby may stop crying for a few seconds and then continue crying. 

I want to hear your thoughts about how to deal with these gaps in time, where an event is still `happening`, as indicated by its absolute start and end times, but momentarily not sounding.

Should I include another attribute in my `SpatialEvents` class that includes these gaps? that could be a way to deal with this level of granularity in the information. 

Please let me know your thoughts @magdalenafuentes and @justinsalamon

**Comment by magdalenafuentes** [comment]
Hey @iranroman thanks for sharing this. I would say you can use the `start` and `end` of `Events`/`SpatialEvents` and clarify in the dataset description that this happens for some annotations. The philosophy we try to always stick to is to not make any modification to the dataset. So in this case, IIUC, a user that's trying to evaluate a model with the `timestamps` in the annotations would run in this problem anyway. So I guess that's ok to include it like that here. Thoughts?

**Comment by iranroman** [comment]
Hello @magdalenafuentes. Thanks for your thoughts. 

This dataset has timestamps for each 100ms of data, so it does not just have a `[start_time, end_time]` for a given event. 

I have implemented my idea of including `silent_times` from each event. To clarify, these `silent times` are part of the annotations. Maybe the way I explained it implied that I was processing the data to identify these `silent_times`, but that is not the case. The dataset annotations explicitly indicate all the `silent_times` for a given event. 

My solution, which includes these `silent_times` in addition of the `[star_time, end_time]`, truly reflects the information in the original dataset annotations, while also adapting the annotation format to what we have been using in `soundata`. 

I'm close to being done with the loader. You'll see my work in a PR in the next few days. If you are curious, here is my fork+branch with my work in progress: https://github.com/iranroman/soundata/tree/tau_nigens_sse_2021

**Comment by justinsalamon** [comment]
Nice progress @iranroman!

@iranroman @magdalenafuentes excuse typos, writing on mobile.

IIUC the events are annotated as time series of active/silent with a 100ms hop size. Is that correct?

We are already supporting time series in SpatialEvents via the arrays of eg azimuth, so I guess this would be one more. That said, this one is about activation which is more general to events and we might want to make sure the API works nicely with other event types in the future.

My 2c would be to call it `activation` (matches eg the A in VAD and the terminology in medleyDB) and store 1’s for when the event is on, rather than represent the silences.

As with the other time series there should be a way to indicate the hop size as well as validate that the time series is as long as the event duration. If the hop size of all the time series (activation, azimuth, etc) is the same that should be made clear and explicit. If it isn’t we need to ensure the series match in duration. Users may also need converter functions to get activation with the same hop as eg azimuth?

Thoughts?

**Comment by iranroman** [comment]
> IIUC the events are annotated as time series of active/silent with a 100ms hop size. Is that correct?

That is correct. In the original dataset annotations, the following are indicated every 100ms (so converter functions will not be needed if we keep all information from the original data): 

* whether a specific sound source is active
* Its azimuth
* Its elevation

Please refer to this link if you want to see how the annotations work and look like in the original dataset: http://dcase.community/challenge2021/task-sound-event-localization-and-detection#reference-labels-and-directions-of-arrival

**Comment by magdalenafuentes** [comment]
@justinsalamon I have a few naive questions: what's the difference between having an `activation` field or having the source appearing and disappearing using `intervals`? Is it because is always the same source and using `intervals` doesn't make it explicit, e.g. it could be different sources appearing and reappearing? Does `activation` truly solve this?
